mabwiser/approximate.py

```diff
@@ -0,0 +1,170 @@
+from mabwiser.ucb import _UCB1
+from mabwiser.utils import Arm, _BaseRNG, create_rng
+
+
+class _ApproximateNeighbors(_Neighbors):
+
+    def fit(self, decisions: np.ndarray, rewards: np.ndarray, contexts: np.ndarray = None) -> NoReturn:
+        # Initialize planes
+        self._initialize(contexts.shape[1])
+
+        # Set the historical data for prediction
+        self.decisions = decisions
+        self.contexts = contexts
+
+        # Binarize the rewards if using Thompson Sampling
+        if isinstance(self.lp, _ThompsonSampling) and self.lp.binarizer:
+            self.rewards = self._binarize_ts_rewards(decisions, rewards)
+        else:
+            self.rewards = rewards
+
+        # Fit hashes for each training context
+        self._fit_operation(contexts)
+
+    def partial_fit(self, decisions: np.ndarray, rewards: np.ndarray,
+                    contexts: Optional[np.ndarray] = None) -> NoReturn:
+
+        # Binarize the rewards if using Thompson Sampling
+        if isinstance(self.lp, _ThompsonSampling) and self.lp.binarizer:
+            rewards = self._binarize_ts_rewards(decisions, rewards)
+
+        # Add more historical data for prediction
+        self.decisions = np.concatenate((self.decisions, decisions))
+        self.rewards = np.concatenate((self.rewards, rewards))
+        self.contexts = np.concatenate((self.contexts, contexts))
+
+        # Fit hashes for each training context
+        self._fit_operation(contexts)
+
+    def _get_neighbors(self, row_2d):
+        """Abstract method to be implemented by child classes."""
+        pass
+
+    def _initialize(self, dimensions):
+        """Abstract method to be implemented by child classes."""
+        pass
+
+    def _fit_operation(self, contexts):
+        """Abstract method to be implemented by child classes."""
+        pass
+
+    def _predict_contexts(self, contexts: np.ndarray, is_predict: bool,
+                          seeds: Optional[np.ndarray] = None, start_index: Optional[int] = None) -> List:
+        # Copy learning policy object
+        lp = deepcopy(self.lp)
+
+        # Create an empty list of predictions
+        predictions = [None] * len(contexts)
+
+        # For each row in the given contexts
+        for index, row in enumerate(contexts):
+
+            # Get random generator
+            lp.rng = create_rng(seed=seeds[index])
+
+            # Prepare for hashing
+            row_2d = row[np.newaxis, :]
+            indices = self._get_neighbors(row_2d)
+
+            # Drop duplicates from list of neighbors
+            indices = list(set(indices))
+
+            # If neighbors exist
+            if len(indices) > 0:
+                predictions[index] = self._get_nhood_predictions(lp, indices, row_2d, is_predict)
+            else:  # When there are no neighbors
+                predictions[index] = self._get_no_nhood_predictions(lp, is_predict)
+
+        return predictions
+
+
+class _LSHNearest(_ApproximateNeighbors):
```

Comment:
I am kinda following the code, it looks very clean. In parallel, it would be great to have a quick "algorithm" session on how this works from a learning perspective. (could be a journal club, or one of the team meetings etc.)